django/core/files/storage.py

```diff
@@ -165,23 +166,57 @@ class FileSystemStorage(Storage):
     """
 
     def __init__(self, location=None, base_url=None, file_permissions_mode=None,
-            directory_permissions_mode=None):
-        if location is None:
-            location = settings.MEDIA_ROOT
-        self.base_location = location
-        self.location = abspathu(self.base_location)
-        if base_url is None:
-            base_url = settings.MEDIA_URL
-        elif not base_url.endswith('/'):
+                 directory_permissions_mode=None):
+        self._location = location
+        if base_url is not None and not base_url.endswith('/'):
             base_url += '/'
-        self.base_url = base_url
-        self.file_permissions_mode = (
-            file_permissions_mode if file_permissions_mode is not None
-            else settings.FILE_UPLOAD_PERMISSIONS
+        self._base_url = base_url
+        self._file_permissions_mode = file_permissions_mode
+        self._directory_permissions_mode = directory_permissions_mode
+        # Weakly reference our receiver in order to make sure we don't end up
+        # with a dangling reference to self preventing our garbage collection.
+        setting_changed.connect(self._clear_cached_properties, weak=True)
+
+    def _clear_cached_properties(self, setting, **kwargs):
+        """Reset setting based property values."""
+        if setting == 'MEDIA_ROOT':
+            self.__dict__.pop('base_location', None)
+            self.__dict__.pop('location', None)
+        elif setting == 'MEDIA_URL':
+            self.__dict__.pop('base_url', None)
+        elif setting == 'FILE_UPLOAD_PERMISSIONS':
+            self.__dict__.pop('file_permissions_mode', None)
+        elif setting == 'FILE_UPLOAD_DIRECTORY_PERMISSIONS':
+            self.__dict__.pop('directory_permissions_mode', None)
+
+    @cached_property
+    def base_location(self):
+        location = self._location
+        return settings.MEDIA_ROOT if location is None else location
```

Comment:
Maybe a helper method would help eliminate the redundancy of these methods? e.g. `return self._value_or_setting(self._location, settings.MEDIA_ROOT)`